### PR TITLE
Allow link to be edited while selection is collapsed

### LINKED ISSUE #10368: Editing a link causes the link url to be inserted into the text
**To Reproduce**
1. Add a link in a paragraph block and save the post
2. Reload the page
3. Attempt to edit and update the link added in step 1 
4. Observe that the url from the link is inserted into the paragraph text

**Expected behavior**
The url for the link should have been changed and not inserted

**Screenshots**
link-container-bug

**Additional context**
- I did a bisect and this seems to have occurred when #7890 was merged (cc: @iseulde).
- It looks to me like the url inserter is falling back to its behaviour of inserting the url when there's no selection. Potentially the selection is lost when the edit/apply buttons are pressed.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #10368. Adds logic to applyFormat to edit the format around the caret if it is of the same type. This logic is already present in `removeFormat`. Adds e2e test.

## How has this been tested?
See #10368.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Review by noisysocks — APPROVED** [review]
Looks great! 👍

File: packages/rich-text/src/test/apply-format.js [hunk 1]
```diff
@@ -49,4 +51,24 @@ describe( 'applyFormat', () => {
...
+			text: 'one two three',
+			start: 4,
+			end: 4,
+		};
+		const result = applyFormat( deepFreeze( record ), a2 );
+
+		expect( result ).toEqual( expected );
+		expect( result ).not.toBe( record );
+		expect( getSparseArrayLength( result.formats ) ).toBe( 3 );
+	} );
```
Comment (noisysocks) [inline packages/rich-text/src/test/apply-format.js]: What happens if `applyFormat` is called with no existing format and a collapsed selection? Should we have a unit test for this case?

Comment (iseulde) [reply packages/rich-text/src/test/apply-format.js]: Then nothing happens, but sure we can have a test for it.

Comment (gziolo) [reply packages/rich-text/src/test/apply-format.js]: I will do it now and merge PR afterwards 👍

Comment (gziolo) [reply packages/rich-text/src/test/apply-format.js]: I hope this works for you @noisysocks: https://github.com/WordPress/gutenberg/pull/10372/commits/9221a33e3a60535221e6d165d96b1385331411a2

**Pull request merged**